### ISSUE TITLE
Another way of gradient backward backpropagation

### ISSUE THREAD

**Issue by ickma2311** [karpathy/micrograd#81]
Hi @karpathy, I followed your code and created my own implementation. I also made some modifications that I believe enhance the code's clarity and understanding.

When forward(add operation as example):
```
out.backwards.extend([(self, 1), (other, 1)])
```
When backward:
```
    def backward(self):


        for node, partial_derivative in self.backwards:
            node.grad += self.grad * partial_derivative
            node.backward()
            if not isinstance(node, Parameter):
                node.zero_grad()
```
I tested it, and it appears to be working well. On the Iris dataset, I achieved an accuracy of 93%. I believe it could reach 100% if I use a more effective loss function.
My code: https://github.com/ickma/picograd

**Comment by dkgitcode** [comment]
In this implementation, you are calculating the gradients immediately rather than deferring them like Karpathy does. While it may enhance clarity, would this not add unnecessary computation during inference ? You are crossing bridges before you even get to them, and there are instances where you don't even need to cross the bridge (i.e forward prop).

**Comment by ickma2311** [comment]
@dkgitcode yes, you are correct. My implementation is worse on efficiency.

**Comment by cxrtisxl** [comment]
My take: don't do anything moving forward. No need for a topological sort, too.
When backward, the parent node always knows its own gradient, op, children, and can calculate their gradients:
```python
# inside Value class:
def back(self, root=True):
        if root:
            self.grad = 1
        if self.op:
            children = self.children
            match self.op:
                case "+":
                    children[0].grad += self.grad
                    children[1].grad += self.grad
                case "*":
                    children[0].grad += self.grad * children[1].data
                    children[1].grad += self.grad * children[0].data
                case "tanh":
                    children[0].grad += self.grad * (1 - math.tanh(children[0].data)**2)
                case _:
                    raise NotImplementedError
            for child in self.children:
                child.back(root=False)
# --------
# Call:
L.back()
```
I also store children in a list, not in a set. So `y = x + x` will produce 2 children (referencing the same `x` node) for `y` which seems logical to me.
